keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md

```diff
@@ -195,153 +242,387 @@ import (
+in operating the external CCMs for those platforms. Please see the following:
 
-_For GA, this section is required: approvers should be able to confirm the
-previous answers based on experience in the field._
+* AWS - [Getting Started with the External Cloud Controller Manager][awsccm]
+* Azure - [Deploy Cloud Controller Manager][azureccm]
+* OpenStack - [Get started with external openstack-cloud-controller-manager in Kubernetes][openstackccm]
+* vSphere - [CPI - Cloud Provider Interface][vsphereccm]
 
-* **Will enabling / using this feature result in any new API calls?**
+An example of deploying an external CCM on GCE can be seen in the
+`start-cloud-controller-manager` script function from the Kubernetes
+repository:
 
-  No, if anything it will result in reduced API calls in core components.
+* GCE - [kubernetes/cluster/gce/gci/configure-helper.sh][gce]
 
-* **Will enabling / using this feature result in introducing new API types?**
+#### General Guidance
 
-  No.
+There are a few general notes to observe when migrating to external CCMs.
+These notes are collected from
+[Kubernetes issue 120411 - Collect tips for external cloud provider migration][issue120411].
 
-* **Will enabling / using this feature result in any new calls to the cloud
-provider?**
+* `kube-apiserver`
+  * Ensure that the `DisableCloudProviders` feature gate is true, whether by default or explicitly.
+* `kubelet`
+  * Ensure that the `DisableKubeletCloudCredentialProviders` feature gate is true, whether by default or explicitly.
+  * Set the `--image-credential-provider-config` and `--image-credential-provider-bin-dir` flags as appropriate for the cloud provider, see the [kubelet Synopsis documentation][kubelet] for more information.
+* `kube-controller-manager`
+  * Disable the `node-ipam-controller` using the `--controllers` command line flag, see the [kube-controller-manager Synopsis documentation][kcm] for more information.
 
-  No, it will actually remove calls to the cloud provider in all core components.
+### Version Skew Strategy
 
-* **Will enabling / using this feature result in increasing size or count of
-the existing API objects?**
+The external CCMs follow the [same version skew policy][skew] as the kube-controller-manager.
+Providers, and communities, that publish their own CCM are responsible for
+updating those CCMs to follow the version and skew policy.
 
-  No.
+Copied from the [Kubernetes Version Skew Policy documentation][skew]:
 
-* **Will enabling / using this feature result in increasing time taken by any
-operations covered by [existing SLIs/SLOs]?**
+> kube-controller-manager, kube-scheduler, and cloud-controller-manager must not be newer than the kube-apiserver instances they communicate with. They are expected to match the kube-apiserver minor version, but may be up to one minor version older (to allow live upgrades).
 
-  No.
+## Production Readiness Review Questionnaire
```

Comment:
@enj - are your concerns addressed by the recent updates?

I'm somewhat biased because we made this migration for GKE and we managed to make it work, so it's certainly possible and the observability was there.
But I somewhat agree that the safety mechanisms aren't perfect, OTOH I'm not entirely sure we have right foundations to build something with a decent ROI and blocking for the sake of blocking wouldn't make sense.

But I would like to hear your thoughts.

Reply:
Yeah I think what we have here is good enough.  This change is incredibly large and having tests that prove full downgrade would be a large effort (but these things are good to think about and manually test at least once).  Having existing tests pass with the in-tree components swapped out serves as a good heuristic for overall functionality as well.